File: apps/web-admin/src/types/fish.ts
/**
 * 鱼百科相关类型定义
 */

// ========== 分类枚举和配置 ==========

export type FishCategory = 'freshwater' | 'saltwater' | 'brackish'

export type Season = 'spring' | 'summer' | 'fall' | 'winter'

export type ActivityLevel = '低' | '中' | '高'

export interface FishCategoryConfig {
  label: string
  icon: string
  color: string
  description: string
}

export const FISH_CATEGORY_CONFIG: Record<FishCategory, FishCategoryConfig> = {
  freshwater: {
    label: '淡水鱼',
    icon: '🐟',
    color: '#1890ff',
    description: '生活在淡水中的鱼类',
  },
  saltwater: {
    label: '海水鱼',
    icon: '🐠',
    color: '#13c2c2',
    description: '生活在海水中的鱼类',
  },
  brackish: {
    label: '广盐鱼',
    icon: '🐡',
    color: '#722ed1',
    description: '可在淡水和海水中生活的鱼类',
  },
}

// 分类选项列表
export const FISH_CATEGORY_OPTIONS = Object.entries(FISH_CATEGORY_CONFIG).map(
  ([value, config]) => ({
    value: value as FishCategory,
    label: `${config.icon} ${config.label}`,
    ...config,
  })
)

// 季节配置
export interface SeasonConfig {
  label: string
  icon: string
  color: string
}

export const SEASON_CONFIG: Record<Season, SeasonConfig> = {
  spring: { label: '春季', icon: '🌸', color: '#52c41a' },
  summer: { label: '夏季', icon: '☀️', color: '#faad14' },
  fall: { label: '秋季', icon: '🍂', color: '#fa8c16' },
  winter: { label: '冬季', icon: '❄️', color: '#1890ff' },
}

export const SEASON_OPTIONS = Object.entries(SEASON_CONFIG).map(([value, config]) => ({
  value: value as Season,
  label: `${config.icon} ${config.label}`,
  ...config,
}))

// 活跃度配置
export const ACTIVITY_LEVEL_CONFIG: Record<ActivityLevel, { label: string; color: string }> = {
  低: { label: '低', color: '#8c8c8c' },
  中: { label: '中', color: '#faad14' },
  高: { label: '高', color: '#52c41a' },
}

export const ACTIVITY_LEVEL_OPTIONS = Object.entries(ACTIVITY_LEVEL_CONFIG).map(
  ([value, config]) => ({
    value: value as ActivityLevel,
    label: config.label,
    ...config,
  })
)

// ========== 知识库接口 ==========

export interface FishKnowledge {
  id: number
  species_id: number
  topic: string
  content: string
  source?: string
  tags?: string
  created_at?: string
  updated_at?: string
}

export interface FishKnowledgeCreate {
  topic: string
  content: string
  source?: string
  tags?: string
}

export interface FishKnowledgeUpdate {
  topic?: string
  content?: string
  source?: string
  tags?: string
}

// ========== 季节活动接口 ==========

export interface FishSeasonActivity {
  id: number
  species_id: number
  season: Season
  activity_level: ActivityLevel
  best_time?: string
  recommended_lures?: string
  fishing_tips?: string
  created_at?: string
  updated_at?: string
}

export interface FishSeasonActivityCreate {
  season: Season
  activity_level: ActivityLevel
  best_time?: string
  recommended_lures?: string
  fishing_tips?: string
}

export interface FishSeasonActivityUpdate {
  season?: Season
  activity_level?: ActivityLevel
  best_time?: string
  recommended_lures?: string
  fishing_tips?: string
}

// ========== 装备推荐接口 ==========

export interface EquipmentRecommendation {
  recommended_lures?: string[]
  recommended_rigs?: string[]
  recommended_rod_power?: string[]
  recommended_line_lb_min?: number
  recommended_line_lb_max?: number
  lure_difficulty?: string
  fight_intensity?: string
}

// ========== 鱼种接口 ==========

export interface FishSpecies {
  species_id: number
  name_cn: string
  name_en?: string
  scientific_name?: string
  category: FishCategory
  habitat?: string
  description?: string
  image_url?: string
  min_weight?: number
  max_weight?: number
  min_length?: number
  max_length?: number
  created_at?: string
  updated_at?: string
  knowledge: FishKnowledge[]
  season_activity: FishSeasonActivity[]
}

export interface FishSpeciesListItem {
  species_id: number
  name_cn: string
  name_en?: string
  category: FishCategory
  habitat?: string
  knowledge_count: number
  season_count: number
  created_at?: string
}

export interface FishSpeciesListResponse {
  items: FishSpeciesListItem[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

// ========== 请求类型 ==========

export interface FishSpeciesCreateRequest {
  name_cn: string
  name_en?: string
  scientific_name?: string
  category: FishCategory
  habitat?: string
  description?: string
  image_url?: string
  min_weight?: number
  max_weight?: number
  min_length?: number
  max_length?: number
}

export interface FishSpeciesUpdateRequest {
  name_cn?: string
  name_en?: string
  scientific_name?: string
  category?: FishCategory
  habitat?: string
  description?: string
  image_url?: string
  min_weight?: number
  max_weight?: number
  min_length?: number
  max_length?: number
}

export interface FishSpeciesListParams {
  page?: number
  page_size?: number
  category?: FishCategory
  keyword?: string
  habitat?: string
}

// ========== 分类统计 ==========

export interface FishCategoryStats {
  category: string
  count: number
  label: string
  icon: string
  color: string
}

export interface FishCategoryStatsResponse {
  categories: FishCategoryStats[]
  total: number
}

// ========== 初始化数据响应 ==========

export interface FishInitDataResponse {
  created_count: number
  message: string
}

File: apps/api/schemas/fish.py
"""
鱼百科 Schema 定义

定义鱼种、知识库、季节活动相关的请求/响应模型
"""

from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class FishCategoryEnum(str, Enum):
    """鱼类分类枚举"""

    FRESHWATER = "freshwater"  # 淡水鱼
    SALTWATER = "saltwater"  # 海水鱼
    BRACKISH = "brackish"  # 广盐鱼


class SeasonEnum(str, Enum):
    """季节枚举"""

    SPRING = "spring"
    SUMMER = "summer"
    FALL = "fall"
    WINTER = "winter"


class ActivityLevelEnum(str, Enum):
    """活跃度枚举"""

    LOW = "低"
    MEDIUM = "中"
    HIGH = "高"


# ========== FishKnowledge Schemas ==========


class FishKnowledgeBase(BaseModel):
    """知识库基础字段"""

    topic: str = Field(..., min_length=1, max_length=100, description="知识主题")
    content: str = Field(..., min_length=1, description="知识内容")
    source: Optional[str] = Field(None, max_length=200, description="信息来源")
    tags: Optional[str] = Field(None, max_length=200, description="标签(逗号分隔)")


class FishKnowledgeCreate(FishKnowledgeBase):
    """创建知识条目请求"""

    pass


class FishKnowledgeUpdate(BaseModel):
    """更新知识条目请求"""

    topic: Optional[str] = Field(None, min_length=1, max_length=100, description="知识主题")
    content: Optional[str] = Field(None, min_length=1, description="知识内容")
    source: Optional[str] = Field(None, max_length=200, description="信息来源")
    tags: Optional[str] = Field(None, max_length=200, description="标签(逗号分隔)")


class FishKnowledgeResponse(FishKnowledgeBase):
    """知识条目响应"""

    id: int
    species_id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# ========== FishSeasonActivity Schemas ==========


class FishSeasonActivityBase(BaseModel):
    """季节活动基础字段"""

    season: SeasonEnum = Field(..., description="季节")
    activity_level: ActivityLevelEnum = Field(..., description="活跃度")
    best_time: Optional[str] = Field(None, max_length=100, description="最佳钓鱼时间")
    recommended_lures: Optional[str] = Field(None, description="推荐拟饵(逗号分隔)")
    fishing_tips: Optional[str] = Field(None, description="钓鱼技巧")


class FishSeasonActivityCreate(FishSeasonActivityBase):
    """创建季节活动请求"""

    pass


class FishSeasonActivityUpdate(BaseModel):
    """更新季节活动请求"""

    season: Optional[SeasonEnum] = Field(None, description="季节")
    activity_level: Optional[ActivityLevelEnum] = Field(None, description="活跃度")
    best_time: Optional[str] = Field(None, max_length=100, description="最佳钓鱼时间")
    recommended_lures: Optional[str] = Field(None, description="推荐拟饵(逗号分隔)")
    fishing_tips: Optional[str] = Field(None, description="钓鱼技巧")


class FishSeasonActivityResponse(BaseModel):
    """季节活动响应"""

    id: int
    species_id: int
    season: str
    activity_level: str
    best_time: Optional[str] = None
    recommended_lures: Optional[str] = None
    fishing_tips: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# ========== FishSpecies Schemas ==========


class FishSpeciesBase(BaseModel):
    """鱼种基础字段"""

    name_cn: str = Field(..., min_length=1, max_length=100, description="中文名")
    name_en: Optional[str] = Field(None, max_length=100, description="英文名")
    scientific_name: Optional[str] = Field(None, max_length=100, description="学名")
    category: FishCategoryEnum = Field(..., description="鱼类分类")
    habitat: Optional[str] = Field(None, max_length=200, description="栖息环境")
    description: Optional[str] = Field(None, description="物种描述")
    image_url: Optional[str] = Field(None, max_length=500, description="图片URL")
    min_weight: Optional[float] = Field(None, ge=0, description="最小体重(kg)")
    max_weight: Optional[float] = Field(None, ge=0, description="最大体重(kg)")
    min_length: Optional[float] = Field(None, ge=0, description="最小体长(cm)")
    max_length: Optional[float] = Field(None, ge=0, description="最大体长(cm)")


class FishSpeciesCreate(FishSpeciesBase):
    """创建鱼种请求"""

    pass


class FishSpeciesUpdate(BaseModel):
    """更新鱼种请求"""

    name_cn: Optional[str] = Field(None, min_length=1, max_length=100, description="中文名")
    name_en: Optional[str] = Field(None, max_length=100, description="英文名")
    scientific_name: Optional[str] = Field(None, max_length=100, description="学名")
    category: Optional[FishCategoryEnum] = Field(None, description="鱼类分类")
    habitat: Optional[str] = Field(None, max_length=200, description="栖息环境")
    description: Optional[str] = Field(None, description="物种描述")
    image_url: Optional[str] = Field(None, max_length=500, description="图片URL")
    min_weight: Optional[float] = Field(None, ge=0, description="最小体重(kg)")
    max_weight: Optional[float] = Field(None, ge=0, description="最大体重(kg)")
    min_length: Optional[float] = Field(None, ge=0, description="最小体长(cm)")
    max_length: Optional[float] = Field(None, ge=0, description="最大体长(cm)")


class FishSpeciesResponse(FishSpeciesBase):
    """鱼种详情响应(含关联数据)"""

    species_id: int
    category: str  # Override to return string instead of enum
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    knowledge: List[FishKnowledgeResponse] = []
    season_activity: List[FishSeasonActivityResponse] = []

    model_config = ConfigDict(from_attributes=True)


class FishSpeciesListItem(BaseModel):
    """鱼种列表项"""

    species_id: int
    name_cn: str
    name_en: Optional[str] = None
    category: str
    habitat: Optional[str] = None
    knowledge_count: int = 0
    season_count: int = 0
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class FishSpeciesListResponse(BaseModel):
    """鱼种分页列表响应"""

    items: List[FishSpeciesListItem]
    total: int
    page: int
    page_size: int
    total_pages: int


# ========== Category Stats ==========


class CategoryStats(BaseModel):
    """分类统计"""

    category: str
    count: int
    label: str
    icon: str
    color: str


class CategoryStatsResponse(BaseModel):
    """分类统计响应"""

    categories: List[CategoryStats]
    total: int


# ========== Equipment Recommendation (Read-only) ==========


class EquipmentRecommendation(BaseModel):
    """装备推荐(只读)"""

    recommended_lures: Optional[List[str]] = Field(None, description="推荐拟饵")
    recommended_rigs: Optional[List[str]] = Field(None, description="推荐钓组")
    recommended_rod_power: Optional[List[str]] = Field(None, description="推荐竿力")
    recommended_line_lb_min: Optional[int] = Field(None, description="推荐鱼线最小拉力")
    recommended_line_lb_max: Optional[int] = Field(None, description="推荐鱼线最大拉力")
    lure_difficulty: Optional[str] = Field(None, description="路亚难度")
    fight_intensity: Optional[str] = Field(None, description="搏斗强度")


# ========== Init Data Response ==========


class InitDataResponse(BaseModel):
    """初始化数据响应"""

    created_count: int
    message: str
